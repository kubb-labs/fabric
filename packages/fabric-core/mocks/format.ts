import type { Options } from 'prettier'
import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string): Promise<string> {
  if (!source) {
    return Promise.resolve('')
  }

  try {
    return prettierFormat(source, formatOptions)
  } catch (_e) {
    return Promise.resolve(source)
  }
}
