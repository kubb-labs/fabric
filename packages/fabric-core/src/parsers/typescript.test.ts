import type ts from 'typescript'
import { format } from '../../mocks/format.ts'

import { print, createImport, createExport } from './typescript.ts'

const formatTS = (elements: ts.Node | (ts.Node | undefined)[]) => {
  return format(print([elements].flat().filter(Boolean)))
}

describe('TypeScript parser', () => {
  test('createImport', async () => {
    expect(
      await formatTS(
        createImport({
          name: 'hello',
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImport({
          name: 'hello',
          path: './hello.ts',
          isTypeOnly: true,
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImport({
          name: ['hello'],
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImport({
          name: 'hello',
          path: './hello.ts',
          isNameSpace: true,
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImport({
          name: [{ propertyName: 'hello', name: 'helloWorld' }],
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()
  })

  test('createExport', async () => {
    expect(
      await formatTS(
        createExport({
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createExport({
          name: ['hello', 'world'],
          asAlias: true,
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createExport({
          name: 'hello',
          asAlias: true,
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    try {
      await formatTS(
        createExport({
          name: 'hello',
          path: './hello.ts',
        }),
      )
    } catch (e) {
      expect(e).toBeDefined()
    }
  })
})
