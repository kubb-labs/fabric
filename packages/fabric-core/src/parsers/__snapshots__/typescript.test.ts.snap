// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TypeScript parser > createExport 1`] = `
"export * from './hello.ts'
"
`;

exports[`TypeScript parser > createExport 2`] = `
"export { hello, world } from './hello.ts'
"
`;

exports[`TypeScript parser > createExport 3`] = `
"export * as hello from './hello.ts'
"
`;

exports[`TypeScript parser > createImport 1`] = `
"import hello from './hello.ts'
"
`;

exports[`TypeScript parser > createImport 2`] = `
"import type hello from './hello.ts'
"
`;

exports[`TypeScript parser > createImport 3`] = `
"import { hello } from './hello.ts'
"
`;

exports[`TypeScript parser > createImport 4`] = `
"import * as hello from './hello.ts'
"
`;

exports[`TypeScript parser > createImport 5`] = `
"import { hello as helloWorld } from './hello.ts'
"
`;
